doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -926,14 +909,257 @@ the conversion in hypotheses :n:`{+ @ident}`.
+
+   Declares a short name for the reduction expression :n:`@red_expr`, for
+   instance ``lazy beta delta [foo bar]``. This short name can then be used
+   in :n:`Eval @ident in` or ``eval`` constructs. This command
+   accepts the :attr:`local` attribute, which indicates that the reduction
+   will be discarded at the end of the
+   file or module. The name is not qualified. In
+   particular declaring the same name in several modules or in several
+   functor applications will be rejected if these declarations are not
+   local. The name :n:`@ident` cannot be used directly as an Ltac tactic, but
+   nothing prevents the user from also performing a
+   :n:`Ltac @ident := @red_expr`.
+
+   .. seealso:: :ref:`applyingconversionrules`
+
+.. todo PR
+   move rewriting.rst to equality.rst once the text is final,
+   perhaps in a different PR to keep the history somewhat understandable
+
+Equality
+--------
+
+.. tacn:: f_equal
+
+   For a goal with the form :n:`f a__1 ... a__n = g b__1 ... b__n`, creates
+   subgoals :n:`f = g` and :n:`a__i = b__i` for the `n` arguments. SImple subgoals, such as those
+   provable by :tacn:`reflexivity` or :tacn:`congruence` are solved automatically.
+
+.. tacn:: reflexivity
+
+   For a goal with the form `t=u`, verifies that `t` and `u` are :term:`convertible`,
+   and if so,
+   solves the goal.  If not, it fails.  It is equivalent to ``apply refl_equal``.
+
+   .. todo PR: this doesn't cover how reflexibity can solve the goal "True"
+
+      also for te goal "1 + 2 = 4", this generates "Unable to unify ..." with a stack
+      trace.  Should have a better message, right?
+
+   .. exn:: The conclusion is not a substitutive equation.
+      :undocumented:
+
+   .. exn:: Unable to unify ... with ...
+      :undocumented:
+
+.. tacn:: symmetry {? in @goal_occurrences }
+
+   Changes a goal that has the form :g:`t = u` into :g:`u = t`.  :n:`@goal_occurrences`
+   may be used to make the change in the selected hypotheses and/or the conclusion.
+   The `at` clause of :n:`@goal_occurrences` is not supported.
+
+.. tacn:: transitivity @one_term
+
+   Changes a goal that has the form :g:`t=u`
+   into the two subgoals :n:`t=@one_term` and :n:`@one_term=u`.
+
+   .. tacn:: etransitivity
+
+      This tactic behaves like :tacn:`transitivity`, using a fresh evar instead of
+      a concrete :token:`one_term`.
+
+Equality and inductive sets
+---------------------------
+
+We describe in this section some special purpose tactics dealing with
+equality and inductive sets or types. These tactics use the
+equality `eq:forall (A:Type), A -> A -> Prop`, simply written with the infix
+symbol `=`.
+
+.. tacn:: decide equality
+
+   Solves a goal of the form :g:`forall x y : R, {x = y} + {~ x = y}`,
+   where :g:`R` is an inductive type such that its constructors do not take
+   proofs or functions as arguments, nor objects in dependent types. It
+   solves goals of the form :g:`{x = y} + {~ x = y}` as well.
```

Comment:
Also `forall y, , {x = y} + {~ x = y}`, and also `or` instead of `sumbool` (`{...} + {...}` is a notation for `sumbool`).  (However, neither `sum` nor `sumor` are supported.)

Reply:
I don't understand what change you're suggesting, not do I understand why you're referring to `sumbool`, `sum` and `sumor`.

Reply:
Here's some relevant context:
- `x <> y` is notation for `~ x = y`
- the `Set Printing All` form of `forall x y : R, {x = y} + {~ x = y}` is `forall x y : R, sumbool (@eq R x y) (not (@eq R x y))`

I'm referring to the fact that `decide equality` will solve the following goals:
```coq
forall x y : nat, {x = y} + {x <> y} (* forall x y : nat, sumbool (@eq nat x y) (not (@eq nat x y)) *)
forall x y : nat, x = y \/ x <> y (* forall x y : nat, or (@eq nat x y) (not (@eq nat x y)) *)
forall y : nat, {x = y} + {x <> y}
forall y : nat, x = y \/ x <> y
{x = y} + {x <> y}
x = y \/ x <> y
```
but does not solve
```coq
forall x y : nat, (x = y) + {x <> y} (* forall x y : nat, sumor (@eq nat x y) (not (@eq nat x y)) *)
forall x y : nat, (x = y) + (x <> y) (* forall x y : nat, sum (@eq nat x y) (not (@eq nat x y)) *)
```

Other interesting facts:

`x` and `y` need not be variables (`decide equality` will solve `1 = 0 \/ 1 <> 0`).

`decide equality` will also solve the following goals:
```coq
forall x y : nat, {x <> y} + {x = y}
forall x y : nat, x <> y \/ x = y
forall x y : nat, {x = y} + {y <> x}
forall x y : nat, x = y \/ y <> x
forall x y : nat, {y <> x} + {y = x}
forall x y : nat, y <> x \/ x = y
forall x y : nat, {y = x} + {y <> x}
forall x y : nat, y = x \/ y <> x
forall x y : nat, {y <> x} + {y = x}
forall x y : nat, y <> x \/ y = x
```
but not
```coq
forall x y : nat, {y = x} + {x <> y}
forall x y : nat, y = x \/ x <> y
forall x y : nat, {x <> y} + {y = x}
forall x y : nat, x <> y \/ y = x
```

Reply:
Thanks.  I tweaked a bit.